core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -576,13 +587,113 @@ private static boolean representAsInternalType(Type type) {
    * Handle decimal type specifically with explicit type conversion
    */
   private static Expression convertAssignableType(
-      Expression argument,  Type targetType) {
+      Expression argument, Type targetType) {
     if (targetType != BigDecimal.class) {
       return argument;
     }
     return convert(argument, targetType);
   }
 
+  public static MethodCallExpression call(Class clazz, String methodName,
+      List<? extends Expression> arguments) {
+    return call(clazz, methodName, arguments, null);
+  }
+
+  /**
+   * The powerful version of {@code org.apache.calcite.linq4j.tree.Expressions#call(
+   * Type, String, Iterable<? extends Expression>)}. Try best effort to convert the
+   * accepted arguments to match parameter type.
+   *
+   * @param clazz Class against which method is invoked
+   * @param methodName Name of method
+   * @param arguments argument expressions
+   * @param targetExpression target expression
+   *
+   * @return MethodCallExpression that call the given name method
+   * @throws RuntimeException if no suitable method found
+   */
+  public static MethodCallExpression call(Class clazz, String methodName,
+       List<? extends Expression> arguments, Expression targetExpression) {
+    Class[] argumentTypes = Types.toClassArray(arguments);
+    try {
+      Method candidate = clazz.getMethod(methodName, argumentTypes);
+      return Expressions.call(targetExpression, candidate, arguments);
+    } catch (NoSuchMethodException e) {
+      for (Method method : clazz.getMethods()) {
+        if (method.getName().equals(methodName)) {
+          final boolean varArgs = method.isVarArgs();
```

Comment:
Is there a need to test `var args` case? (not sure if it's already being tested in codebase somewhere else)

Reply:
Current test in the PR for `SqlFunction.array(Object... args)` already covers this path. Moreover, the logic is borrowed from `Types.lookupMethod` in `Linq4j`, I think it can be trusted.